src/main/java/randoop/test/ContractCheckingVisitor.java

```diff
@@ -99,6 +101,71 @@ public TestChecks visit(ExecutableSequence s) {
   }
 
   /**
+   * Checks a ternary contract over the set of values defined in the sequence,
+   * and attaches failing checks at final statement of the sequence.
+   *
+   * @param s
+   *          the executable sequence
+   * @param c
+   *          the contract to check
+   * @param values
+   *          the set of positions defining values to check
+   * @param checks
+   *          the {@code TestChecks} to which new checks are added
+   */
+  private void checkTernary(
+          ExecutableSequence s, ObjectContract c, Set<Integer> values, ErrorRevealingChecks checks) {
+    for (Integer i : values) {
+      for (Integer j : values) {
+        for (Integer k : values) {
+          // Create three ExecutionOutcome objects from the ExecutableSequence
+          ExecutionOutcome result1 = s.getResult(i);
+          ExecutionOutcome result2 = s.getResult(j);
+          ExecutionOutcome result3 = s.getResult(k);
+
+          if (Log.isLoggingOn()) {
+            Log.logLine("Checking contract " + c.getClass() + " on " + i + ", " + j + ", " + k);
+          }
+
+          ExecutionOutcome exprOutcome =
+                  ObjectContractUtils.execute(
+                          c,
+                          ((NormalExecution) result1).getRuntimeValue(),
+                          ((NormalExecution) result2).getRuntimeValue(),
+                          ((NormalExecution) result3).getRuntimeValue());
+
+          if (exprOutcome instanceof NormalExecution) {
+            NormalExecution e = (NormalExecution) exprOutcome;
+            if (e.getRuntimeValue().equals(true)) {
+              if (Log.isLoggingOn()) Log.logLine("Contract returned true.");
+              // Behavior ok.
+            } else {
+              if (Log.isLoggingOn())
+                Log.logLine("Contract returned false. Will add ExpressionEqFalse check");
+              // Create an check that records the actual value
+              // returned by the expression, marking it as invalid
+              // behavior.
+              checks.add(new ObjectCheck(c, i, s.sequence.getVariable(i), s.sequence.getVariable(j), s.sequence.getVariable(k)));
+            }
+          } else if (exprOutcome instanceof ExceptionalExecution) {
+            Throwable e = ((ExceptionalExecution) exprOutcome).getException();
+            if (Log.isLoggingOn()) Log.logLine("Contract threw exception: " + e.getMessage());
+            if (e instanceof BugInRandoopException) {
+              throw new BugInRandoopException(e);
+            }
+            // Execution of contract resulted in exception. Do not create
+            // a contract-violation decoration.
```

Comment:
Randoop has not been doing anything with exceptions thrown by contracts
that are not Randoop errors.  This is probably because the contract is not
part of the test method sequence, and exceptions from statements in a
sequence are dealt with by a different visitor.





On Mon, May 30, 2016 at 8:23 PM, Michael Ernst <notifications@github.com>
wrote:

> In src/main/java/randoop/test/ContractCheckingVisitor.java
> <https://github.com/randoop/randoop/pull/99#discussion_r65119673>:
>
> > +            } else {
> > +              if (Log.isLoggingOn())
> > +                Log.logLine("Contract returned false. Will add ExpressionEqFalse check");
> > +              // Create an check that records the actual value
> > +              // returned by the expression, marking it as invalid
> > +              // behavior.
> > +              checks.add(new ObjectCheck(c, i, s.sequence.getVariable(i), s.sequence.getVariable(j), s.sequence.getVariable(k)));
> > +            }
> > +          } else if (exprOutcome instanceof ExceptionalExecution) {
> > +            Throwable e = ((ExceptionalExecution) exprOutcome).getException();
> > +            if (Log.isLoggingOn()) Log.logLine("Contract threw exception: " + e.getMessage());
> > +            if (e instanceof BugInRandoopException) {
> > +              throw new BugInRandoopException(e);
> > +            }
> > +            // Execution of contract resulted in exception. Do not create
> > +            // a contract-violation decoration.
>
> I don't think this is just when e.getRuntimeValue().equals(true)
> evaluates to false, but I'm not sure.
>
> I'm trying to understand under what circumstances execution reaches line
> 156, which seems to be whenever an exception other than
> BugInRandoopException is thrown. The code seems to silently ignore all
> such exceptions.
>
> Ignoring an exception is a code smell and I wonder what is the
> justification for ignoring the exception.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/randoop/randoop/pull/99/files/3e0cffc1ccd68c0fc501fae40637c3701e7fa66b#r65119673>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFOdvXiW51zmtQFd6TvvmeHp7J1C8INhks5qG6nCgaJpZM4Ij6Nm>
> .
>



-- 
Ben Keller
bjkeller@uw.edu